File: 2307-replace-non-coprime-numbers-in-array/replace-non-coprime-numbers-in-array.java
class Solution {
    public List<Integer> replaceNonCoprimes(int[] nums) {
        List<Integer> res = new ArrayList<>();

        for (int num : nums) {
            // Merge while last element and current num are non-coprime
            while (!res.isEmpty() && gcd(res.get(res.size() - 1), num) > 1) {
                num = lcm(res.get(res.size() - 1), num);
                res.remove(res.size() - 1); // pop last
            }
            res.add(num);
        }

        return res;
    }

    // Helper to compute GCD
    private int gcd(int a, int b) {
        while (b != 0) {
            int tmp = a % b;
            a = b;
            b = tmp;
        }
        
        return a;
    }

    // Helper to compute LCM
    private int lcm(int a, int b) {
        return a / gcd(a, b) * b;
    }
}

File: 2307-replace-non-coprime-numbers-in-array/replace-non-coprime-numbers-in-array.py
class Solution:
    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        res = []  # This will store the final result array

        # Loop through each number in the input list
        for num in nums:
            # While the result array is not empty AND the last element in res
            # has GCD > 1 with the current num (i.e., they are non-coprime)
            while res and gcd(res[-1], num) > 1:
                # Replace num with the LCM of res[-1] and num
                # Because the two adjacent non-coprime numbers should merge into one
                num = lcm(res[-1], num)

                # Remove the last element (since it merged into num)
                res.pop()

            # Push the current (possibly merged) number into res
            res.append(num)

        # Return the final modified array
        return res
